python/ray/data/_internal/streaming_repartition.py

```diff
@@ -0,0 +1,274 @@
+import math
+from collections import deque
+from dataclasses import dataclass
+from typing import Deque, Iterable, List, Optional
+
+from ray.data._internal.delegating_block_builder import DelegatingBlockBuilder
+from ray.data._internal.execution.interfaces import RefBundle, TaskContext
+from ray.data._internal.execution.operators.map_operator import _TaskInput
+from ray.data.block import Block, BlockAccessor, BlockMetadata, Schema
+from ray.types import ObjectRef
+
+"""Streaming repartition builds fixed-size outputs from a stream of inputs.
+
+    We construct batches here to produce exactly sized outputs from arbitrary [start, end) slices across input blocks.
+    The task builder submits a map task only after the total number of rows accumulated across pending blocks reaches
+    target num rows (except during the final flush, which may emit a smaller tail block). This allows us to create
+    target-sized batches without materializing entire large blocks on the driver.
+"""
+
+
+# Key used to stash the repartition task spec in the TaskContext kwargs
+# for the downstream execution function to read and act upon.
+STREAMING_REPARTITION_SPEC_KEY = "streaming_repartition_spec"
+
+
+@dataclass
+class StreamingRepartitionContributorSpec:
+    """A single slice contribution to an output block.
+
+    Fields:
+    - block_index: Index of the source block within the current task input bundle.
+    - start_offset/end_offset: Half-open row range [start, end) to take from that
+      source block.
+    """
+
+    # Index of the contributing block within the task input bundle.
+    block_index: int
+    # Slice [start_offset, end_offset) in the contributing block.
+    start_offset: int
+    end_offset: int
+
+    @property
+    def slice_rows(self) -> int:
+        """Number of rows contributed by this slice."""
+        return self.end_offset - self.start_offset
```

Comment:
This name suggests to me that it slices the rows, rather than providing a row count. A name like `num_rows` or `num_rows_in_slice` might be clearer